### DISCUSSION TITLE
use a resource that uses `api.url` allow origin in api's allow.origin

### DISCUSSION THREAD

**Discussion by tsuf239** [winglang/wing#6307] [Announcements]
### I tried this:

```js
bring cloud;

class WebsiteServer {
  pub url: str;
  new (env: Map<str>) {
    // initializing some terraform and stuff....
    // doing some stuff with the env.
    this.url = "some terraform token";
  }
}

// creating a url that allows only the server origin
let api = new cloud.Api(corsOptions:{allowOrigin: server.url});

// the website server resource needs the api.url in its environment- since it's 
// essential to the initiation.
let server = new WebsiteServer({API_URL: api.url});
```
### This happened:

I couldn't use a property `url` of a resource that wasn't initialized yet :(

### I expected this:

to have a way to pass that

### Wing Version

0.66.5

### Community Notes

* Please vote by adding a 👍 reaction to the issue to help us prioritize.
* If you are interested to work on this issue, please leave a comment.

**Comment by tsuf239** [comment]
[example in the playground](https://www.winglang.io/play/?code=YgByAGkAbgBnACAAYwBsAG8AdQBkADsACgAKAGMAbABhAHMAcwAgAFcAZQBiAHMAaQB0AGUAUwBlAHIAdgBlAHIAIAB7AAoAIAAgAHAAdQBiACAAdQByAGwAOgAgAHMAdAByADsACgAgACAAbgBlAHcAIAAoAGUAbgB2ADoAIABNAGEAcAA8AHMAdAByAD4AKQAgAHsACgAgACAAIAAgAC8ALwAgAGkAbgBpAHQAaQBhAGwAaQB6AGkAbgBnACAAcwBvAG0AZQAgAHQAZQByAHIAYQBmAG8AcgBtACAAYQBuAGQAIABzAHQAdQBmAGYALgAuAC4ALgAKACAAIAAgACAALwAvACAAZABvAGkAbgBnACAAcwBvAG0AZQAgAHMAdAB1AGYAZgAgAHcAaQB0AGgAIAB0AGgAZQAgAGUAbgB2AC4ACgAgACAAIAAgAHQAaABpAHMALgB1AHIAbAAgAD0AIAAiAHMAbwBtAGUAIAB0AGUAcgByAGEAZgBvAHIAbQAgAHQAbwBrAGUAbgAiADsACgAgACAAfQAKAH0ACgAKAC8ALwAgAGMAcgBlAGEAdABpAG4AZwAgAGEAIAB1AHIAbAAgAHQAaABhAHQAIABhAGwAbABvAHcAcwAgAG8AbgBsAHkAIAB0AGgAZQAgAHMAZQByAHYAZQByACAAbwByAGkAZwBpAG4ACgBsAGUAdAAgAGEAcABpACAAPQAgAG4AZQB3ACAAYwBsAG8AdQBkAC4AQQBwAGkAKABjAG8AcgBzAE8AcAB0AGkAbwBuAHMAOgB7AGEAbABsAG8AdwBPAHIAaQBnAGkAbgA6ACAAcwBlAHIAdgBlAHIALgB1AHIAbAB9ACkAOwAKAAoALwAvACAAdABoAGUAIAB3AGUAYgBzAGkAdABlACAAcwBlAHIAdgBlAHIAIAByAGUAcwBvAHUAcgBjAGUAIABuAGUAZQBkAHMAIAB0AGgAZQAgAGEAcABpAC4AdQByAGwAIABpAG4AIABpAHQAcwAgAGUAbgB2AGkAcgBvAG4AbQBlAG4AdAAtACAAcwBpAG4AYwBlACAAaQB0ACcAcwAgAAoALwAvACAAZQBzAHMAZQBuAHQAaQBhAGwAIAB0AG8AIAB0AGgAZQAgAGkAbgBpAHQAaQBhAHQAaQBvAG4ALgAKAGwAZQB0ACAAcwBlAHIAdgBlAHIAIAA9ACAAbgBlAHcAIABXAGUAYgBzAGkAdABlAFMAZQByAHYAZQByACgAewBBAFAASQBfAFUAUgBMADoAIABhAHAAaQAuAHUAcgBsAH0AKQA7AAoACgA%3D)

**Comment by Chriscbr** [comment]
> [example in the playground](https://www.winglang.io/play/?code=YgByAGkAbgBnACAAYwBsAG8AdQBkADsACgAKAGMAbABhAHMAcwAgAEkAbQBhAGcAaQBuAGEAcgB5AFIAZQBzAG8AdQByAGMAZQAgAHsACgAgACAAcAB1AGIAIAB1AHIAbAA6ACAAcwB0AHIAOwAKACAAIABuAGUAdwAgACgAZQBuAHYAOgAgAE0AYQBwADwAcwB0AHIAPgApACAAewAKACAAIAAgACAALwAvACAAaQBuAGkAdABpAGEAbABpAHoAaQBuAGcAIABzAG8AbQBlACAAdABlAHIAcgBhAGYAbwByAG0AIABhAG4AZAAgAHMAdAB1AGYAZgAuAC4ALgAuAAoAIAAgACAAIAAvAC8AIABkAG8AaQBuAGcAIABzAG8AbQBlACAAcwB0AHUAZgBmACAAdwBpAHQAaAAgAHQAaABlACAAZQBuAHYALgAKACAAIAAgACAAdABoAGkAcwAuAHUAcgBsACAAPQAgACIAcwBvAG0AZQAgAHQAZQByAHIAYQBmAG8AcgBtACAAdABvAGsAZQBuACIAOwAKACAAIAB9AAoAfQAKAAoALwAvACAAYwByAGUAYQB0AGkAbgBnACAAYQAgAHUAcgBsACAAdABoAGEAdAAgAHMAdQBwAHAAbwByAHQAcwAgAG8AbgBsAHkAIAB0AGgAZQAgAGkAbQBhAGcAaQBuAGEAcgB5ACAAbwByAGkAZwBpAG4ACgBsAGUAdAAgAGEAcABpACAAPQAgAG4AZQB3ACAAYwBsAG8AdQBkAC4AQQBwAGkAKABjAG8AcgBzAE8AcAB0AGkAbwBuAHMAOgB7AGEAbABsAG8AdwBPAHIAaQBnAGkAbgA6ACAAaQBtAGEAZwBpAG4AYQByAHkALgB1AHIAbAB9ACkAOwAKAAoALwAvACAAdABoAGUAIABpAG0AYQBnAGkAbgBhAHIAeQAgAHIAZQBzAG8AdQByAGMAZQAgAG4AZQBlAGQAcwAgAHQAaABlACAAYQBwAGkALgB1AHIAbAAgAGkAbgAgAGkAdABzACAAZQBuAHYAaQByAG8AbgBtAGUAbgB0AC0AIABzAGkAbgBjAGUAIABpAHQAJwBzACAACgAvAC8AIABlAHMAcwBlAG4AdABpAGEAbAAgAHQAbwAgAHQAaABlACAAaQBuAGkAdABpAGEAdABpAG8AbgAuAAoAbABlAHQAIABpAG0AYQBnAGkAbgBhAHIAeQAgAD0AIABuAGUAdwAgAEkAbQBhAGcAaQBuAGEAcgB5AFIAZQBzAG8AdQByAGMAZQAoAHsAQQBQAEkAXwBVAFIATAA6ACAAYQBwAGkALgB1AHIAbAB9ACkAOwAKAAoACgA%3D)

It looks like there's a dependency cycle between the resources in the code (each is trying to depend on the other). Is it possible to make one of them add configuration later? e.g.

```js
bring cloud;

class ImaginaryResource {
  pub url: str;
  env: MutMap<str>;
  new () {
    // initializing some terraform and stuff....
    // doing some stuff with the env.
    this.url = "some terraform token";
    this.env = MutMap<str>{};
  }
  pub addEnv(key: str, value: str) {
    this.env.set(key, value);
  }
}

// the imaginary resource needs the api.url in its environment- since it's 
// essential to the initiation.
let imaginary = new ImaginaryResource();

// creating a url that supports only the imaginary origin
let api = new cloud.Api(corsOptions:{allowOrigin: imaginary.url});

imaginary.addEnv("API_URL", api.url);
```

(some way to add CORS options to `cloud.Api` after it is constructed would also be ok I think)

**Comment by tsuf239** [comment]
also, this may work:
```
bring cloud;

class WebsiteServer {
  pub url: str;
  new (env: Map<str>) {
    // initializing some terraform and stuff....
    // doing some stuff with the env.
    this.url = "some terraform token";
  }
}
let var server: WebsiteServer? = nil;

// creating a url that allows only the server origin
let api = new cloud.Api(corsOptions:{allowOrigin: server?.url});

// the website server resource needs the api.url in its environment- since it's 
// essential to the initiation.
server = new WebsiteServer({API_URL: api.url});
```
[playground here](https://www.winglang.io/play/?code=YgByAGkAbgBnACAAYwBsAG8AdQBkADsACgAKAGMAbABhAHMAcwAgAFcAZQBiAHMAaQB0AGUAUwBlAHIAdgBlAHIAIAB7AAoAIAAgAHAAdQBiACAAdQByAGwAOgAgAHMAdAByADsACgAgACAAbgBlAHcAIAAoAGUAbgB2ADoAIABNAGEAcAA8AHMAdAByAD4AKQAgAHsACgAgACAAIAAgAC8ALwAgAGkAbgBpAHQAaQBhAGwAaQB6AGkAbgBnACAAcwBvAG0AZQAgAHQAZQByAHIAYQBmAG8AcgBtACAAYQBuAGQAIABzAHQAdQBmAGYALgAuAC4ALgAKACAAIAAgACAALwAvACAAZABvAGkAbgBnACAAcwBvAG0AZQAgAHMAdAB1AGYAZgAgAHcAaQB0AGgAIAB0AGgAZQAgAGUAbgB2AC4ACgAgACAAIAAgAHQAaABpAHMALgB1AHIAbAAgAD0AIAAiAHMAbwBtAGUAIAB0AGUAcgByAGEAZgBvAHIAbQAgAHQAbwBrAGUAbgAiADsACgAgACAAfQAKAH0ACgBsAGUAdAAgAHYAYQByACAAcwBlAHIAdgBlAHIAOgAgAFcAZQBiAHMAaQB0AGUAUwBlAHIAdgBlAHIAPwAgAD0AIABuAGkAbAA7AAoACgAvAC8AIABjAHIAZQBhAHQAaQBuAGcAIABhACAAdQByAGwAIAB0AGgAYQB0ACAAYQBsAGwAbwB3AHMAIABvAG4AbAB5ACAAdABoAGUAIABzAGUAcgB2AGUAcgAgAG8AcgBpAGcAaQBuAAoAbABlAHQAIABhAHAAaQAgAD0AIABuAGUAdwAgAGMAbABvAHUAZAAuAEEAcABpACgAYwBvAHIAcwBPAHAAdABpAG8AbgBzADoAewBhAGwAbABvAHcATwByAGkAZwBpAG4AOgAgAHMAZQByAHYAZQByAD8ALgB1AHIAbAB9ACkAOwAKAAoALwAvACAAdABoAGUAIAB3AGUAYgBzAGkAdABlACAAcwBlAHIAdgBlAHIAIAByAGUAcwBvAHUAcgBjAGUAIABuAGUAZQBkAHMAIAB0AGgAZQAgAGEAcABpAC4AdQByAGwAIABpAG4AIABpAHQAcwAgAGUAbgB2AGkAcgBvAG4AbQBlAG4AdAAtACAAcwBpAG4AYwBlACAAaQB0ACcAcwAgAAoALwAvACAAZQBzAHMAZQBuAHQAaQBhAGwAIAB0AG8AIAB0AGgAZQAgAGkAbgBpAHQAaQBhAHQAaQBvAG4ALgAKAHMAZQByAHYAZQByACAAPQAgAG4AZQB3ACAAVwBlAGIAcwBpAHQAZQBTAGUAcgB2AGUAcgAoAHsAQQBQAEkAXwBVAFIATAA6ACAAYQBwAGkALgB1AHIAbAB9ACkAOwAKAAoA)

**Comment by Chriscbr** [comment]
Recapping some of the options the team came up with offline:
- If the user's adds a custom domain with `cloud.Domain` as a third resource, then the URL will be known at compile time so it may be possible to avoid the dependency cycle. Pseudocode:

```js
let domain = new cloud.Domain(url: "example.com");
let api = new cloud.Api(corsOptions:{allowOrigin: domain.url});
let server = new WebsiteServer({ API_URL: domain.url });
```

- The user can try to add some late-binding methods to their class so that API_URL does not need to be added in the constructor, like `server.addEnvironment("API_URL", api.url)` (see earlier comment)
- We can add some late-binding methods so that the CORS options can be added later in the preflight code, like `api.addCorsOptions(allowOrigin: server.url)`. An addition to `cloud.Api` would be welcome!
- If the user is compiling the app for Terraform, they can pass a token which resolves at synthesis time to the correct value. Pseudocode:

```js
let var serverUrl: str? = nil;

let api = new cloud.Api(corsOptions:{
  allowOrigin: cdktf.Lazy.string({ produce: () => { return serverUrl!; } }),
});

let server = new WebsiteServer({API_URL: api.url});
serverUrl = server.url;
```

It's possible we'll need to add support for something like this to the `sim` target, or more generally a cloud-agnostic token API to the language / standard library.
